### ISSUE TITLE
Sparkpost Integration Broken 

### ISSUE THREAD

**Issue by aheicher** [r-a-y/bp-reply-by-email#105]
(probably)

Hi r-a-y, as always, thanks for the amazing plugin (and child plugins). It's appreciated.

Wanted to let you know that I *think* a recent Sparkpost change has broken your integration. One of my users alerted me that new topic and reply by email seemed to randomly break around 7/20, 7/21.  No changes had been made to the Wordpress site since 6/10, much prior, and no changes were made to the Sparkpost setup on my end, or the DNS setup with our host. 

The change in the [changelog](https://www.sparkpost.com/docs/tech-resources/change-log-sparkpost/) that I think might have caused it is "Mask sensitive information in email headers" but that is extreme speculation on my part, as I have no idea how your plugin handles that.

Nothing was being written to the log file.

After a lot of fruitless troubleshooting, I finally changed the DNS to Postmark's, created an account with them, configured Postmark and your plugin, and things were back up and working straight away. As far as I'm concerned, the issue is closed, but for the rest of your users, this might cause some problems, assuming my theory is right.

Thanks!

**Comment by r-a-y** [comment]
I'm looking at the POST response by SparkPost and everything looks fine to me.

Nothing has changed.  If you are interested in looking into this, I would suggest looking into debugging SparkPost some more. Follow SparkPost's own debug guide located under the "Troubleshooting and FAQ" section of this page:
https://support.sparkpost.com/customer/portal/articles/2039614-enabling-inbound-email-relaying-relay-webhooks

**Comment by aheicher** [comment]
Suppose I stand corrected. Jeez, no idea what broke, then, but at least it's working with my current setup. I'll look more into debugging with Sparkpost on my staging site and I'll be sure to let you know if it has anything to do with rbe, although at this point I kind of doubt it.

Closing this out. Thanks for the quick response!

**Comment by r-a-y** [comment]
Also, just in case you're using the latest master branch, the debug log has moved from `/wp-content/bp-rbe-debug.log` to `/wp-content/uploads/bp-rbe-debug.log` if that helps or not.

**Comment by boonebgorges** [comment]
Tentatively reopening, as I ran into a similar problem on a client site.

In my case, it appears that SparkPost webhooks are no longer providing the `X-MessageSystems-Webhook-Token` header. At least, the `HTTP_X_MESSAGESYSTEMS_WEBHOOK_TOKEN` key isn't set in `$_SERVER`. Is it possible that SparkPost has changed something, or could it be the webhost (WP Engine, in my case) who's not passing along the header for some reason?

**Comment by r-a-y** [comment]
I tested this a few days ago on a site using SparkPost and RBE was working for me.

The header exists.  You can verify by changing the webhook URL to something like Runscope to verify the POST response and headers.

Might be worth testing WPEngine to see if they accept HTTP X Headers.  If not, I can add a filter to bypass the header check.

**Comment by r-a-y** [comment]
Maybe SparkPost no longer adds the `X-MessageSystems-Webhook-Token` header if you do not add a webhook token when you created the webhook.

When I set up SparkPost, I did add the auth token:
https://github.com/r-a-y/bp-reply-by-email/wiki/SparkPost#5-in-sparkpost-register-the-webhook

Did you, @boonebgorges and @aheicher?

**Comment by aheicher** [comment]
I did not set up an auth token, no.

**Comment by boonebgorges** [comment]
My client set it up, but I don't believe she set that token. That might explain it.

**Comment by r-a-y** [comment]
Thanks for responding, @aheicher.	

I think for security purposes, the `auth_token` parameter should be required from here on out.

I will modify the documentation to make this step mandatory.  Would be nice to verify if adding the auth token will fix the issue.

**Comment by aheicher** [comment]
@r-a-y I can't promise an expedited investigation, but there is a staging site for the client that was experiencing this issue, so if I am able to test it there I will report back. I think we'll actually be sticking with Postmark for usability and reporting reasons, but it'd be good to know if this is your issue.

**Comment by boonebgorges** [comment]
Adding the `auth_token` to the relay webhook appears to have worked! Good thinking, @r-a-y. If you could update the docs, it would be great (and this issue can be closed).

**Comment by r-a-y** [comment]
Thanks for verifying, @boonebgorges!

I've updated the wiki page:
https://github.com/r-a-y/bp-reply-by-email/wiki/SparkPost#5-in-sparkpost-register-the-webhook	

Feel free to make further updates.

I just checked SparkPost's webhook documentation and it appears that they have deprecated the `auth_token` parameter:
https://developers.sparkpost.com/api/webhooks.html#header-webhooks-object-properties

We might have to change this in the future to their newer `custom_headers` parameter.

Going to close this one.  Thanks for looking into this, you two!
